### PR TITLE
A new test to check config attributes being used

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Add a new test to check config attributes being used.

For edge cases, I only add rules to 2 files. If the concept is approved,  **I will add more to pass CI**, and continue the work of cleaning up in follow-up PR(s).

### CODE REVIEW

File: tests/models/biogpt/test_modeling_biogpt.py [hunk 1]
```diff
@@ -273,7 +273,10 @@ def setUp(self):
         self.config_tester = ConfigTester(self, config_class=BioGptConfig, hidden_size=37)
 
     def test_config(self):
-        self.config_tester.run_common_tests()
+        def check_fn(attribute):
+            return attribute == "layer_norm_eps"
```
Comment (ydshieh) [inline tests/models/biogpt/test_modeling_biogpt.py:277]: This is the way to allow some cases where an attribute usage is not found but we know clearly the reason and cannot change anything

File: tests/models/whisper/test_modeling_whisper.py [hunk 2]
```diff
@@ -278,7 +278,11 @@ def setUp(self):
         self.maxDiff = 3000
 
     def test_config(self):
-        self.config_tester.run_common_tests()
+        # TODO: check and potentially remove these unused attributes, then update `check_fn`
```
Comment (ydshieh) [inline tests/models/whisper/test_modeling_whisper.py:281]: This is to skip the failing cases before we find time to check and fix them.

File: tests/models/whisper/test_modeling_whisper.py [hunk 3]
```diff
@@ -278,7 +278,11 @@ def setUp(self):
...
+        def check_fn(attribute):  # noqa
+            return True
```
Comment (ydshieh) [inline tests/models/whisper/test_modeling_whisper.py:283]: This is to skip the failing cases before we find time to check and fix them.

File: tests/test_configuration_common.py [hunk 4]
```diff
@@ -205,7 +207,131 @@ def check_config_arguments_init(self):
...
+
+                # Allow cases without checking the default value in the configuration class
+                elif attribute in attributes_to_allow + attributes_used_in_generation:
+                    case_allowed = True
+                elif attribute.endswith("_token_id"):
+                    case_allowed = True
+
+                # configuration class specific cases
+                if not case_allowed and check_fn is not None:
+                    case_allowed = check_fn(attribute)
```
Comment (ydshieh) [inline tests/test_configuration_common.py:281]: If none of the above **common** checks could not find the usage of an attribute or allow the edge case, we check further with the model specific rules given in `check_fn` that is passed from specific model testing files.

**Comment by ydshieh** [comment]
@sgugger Ready for review - The failing tests will be addressed (by adding specific rules in subclasses) once the PR is approved 🙏 .

**Review by sgugger — COMMENTED** [review]
Thanks for working on this. I don't think a common config test is the right place for this as this is a quality tool, not a correctness tool. So it should be more in the utils folder and integrated in `make repo-consistency` rather than the tests.

For the special cases we cna have a mapping model type to attribute to ignore instead of a special function to overwrite in the testers.

File: utils/check_config_attributes.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+CASES_TO_ALLOW = {
+    "BioGptConfig": ["layer_norm_eps"],
+    "CvtConfig": ["layer_norm_eps"],
+    "DPRConfig": ["layer_norm_eps"],
+    "GLPNConfig": ["layer_norm_eps"],
+    "PerceiverConfig": ["layer_norm_eps"],
+    "RetriBertConfig": ["layer_norm_eps"],
+    "SegformerConfig": ["layer_norm_eps"],
+    "TrajectoryTransformerConfig": ["layer_norm_eps"],
+}
```
Comment (ydshieh) [inline utils/check_config_attributes.py:46]: To add more entries to pass CI (with TODO comments for some cases)

**Comment by ydshieh** [comment]
The report looks like (if unused attributes are detected)
```bash
ValueError: The following configuration classes contain unused attributes in the corresponding modeling files:
CLIPSegConfig: ['decoder_attention_dropout', 'decoder_hidden_act']
...
DinatConfig: ['patch_norm']
...
```

File: utils/check_config_attributes.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                "summary_first_dropout",
+            ]:
+                if "SequenceSummary" in modeling_source:
+                    attribute_used = True
+            if attribute_used:
+                break
+        if attribute_used:
+            break
+
+    # common and important attributes, although not always appear in the modeling files
```
Comment (sgugger) [inline utils/check_config_attributes.py:87]:
```suggestion
    # common and important attributes, even if they do not always appear in the modeling files
```

File: utils/check_config_attributes.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    "GLPNConfig": ["layer_norm_eps"],
+    "PerceiverConfig": ["layer_norm_eps"],
+    "RetriBertConfig": ["layer_norm_eps"],
+    "SegformerConfig": ["layer_norm_eps"],
+    "TrajectoryTransformerConfig": ["layer_norm_eps"],
+}
+
+
+def check_attribute_being_used(config_class, attributes, default_value, modeling_sources):
+    """Check if any name in `attributes` is used in one of the strings in`modeling_sources`"""
```
Comment (sgugger) [inline utils/check_config_attributes.py:50]:
```suggestion
    """Check if any name in `attributes` is used in one of the strings in `modeling_sources`"""
```
Would be nice to also explain `config_class` and `default_value` here :-)

File: utils/check_config_attributes.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                case_allowed = True
+
+            # configuration class specific cases
+            if not case_allowed:
+                case_allowed = attribute in CASES_TO_ALLOW.get(config_class.__name__, [])
+
+    return attribute_used or case_allowed
+
+
+def check_config_attributes_being_used(config_class):
```
Comment (sgugger) [inline utils/check_config_attributes.py:123]: A small docstring would be nice!

File: utils/check_config_attributes.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+            attributes.append(reversed_attribute_map[config_param])
+
+        if not check_attribute_being_used(config_class, attributes, default_value, modeling_sources):
+            unused_attributes.append(attributes[0])
+
+    return sorted(unused_attributes)
+
+
+def check_config_attributes():
+    configs_with_unused_attributes = {}
```
Comment (sgugger) [inline utils/check_config_attributes.py:164]: Same here, a small docstring would be nice :-)

**Review by sgugger — COMMENTED** [review]
Thanks for iterating! Just left a couple more comments.

As you noticed the job is failing right now, are you planning to add everything in the special map or to fix all the attributes not used?

**Comment by ydshieh** [comment]
> Thanks for iterating! Just left a couple more comments.
> 
> As you noticed the job is failing right now, are you planning to add everything in the special map or to fix all the attributes not used?

Yes, that's mentioned in the description/comment (in previous version, it's better doing so, but with this new version you suggested, I can indeed add them earlier though).

FYI: the special map I will update will contain
- some confirmed allowed cases (i.e. we know the reasons and nothing we can do but just allow)
- some skipped cases **for now** to allow temporarily with #TODO comment

**Review by sgugger — APPROVED** [review]
Plan LGTM!

**Pull request merged**